### PR TITLE
fix(element): corrects the gradient enteries

### PR DESCRIPTION
**Pull request by mohitpundir**
Earlier, we computed the gradients as du_j/dx_i (spatial/component) as entries, which is conventionally written as du_i/dx_j (component/spatial) [Reference](https://www.continuummechanics.org/deformationgradient.html), and would have led to an error for asymmetric gradients.

### CODE REVIEW

**Comment by zrlf** [comment]
Further made the multiplication shape agnostic (also works for gradients of tensor fields) by using `einsum`:
- `(dNdX @ values).T` becomes `jnp.einsum("dn,n...->...d", dNdX, values)`
- `N @ values` becomes `jnp.einsum("n,n...->...", N, values)`

Added the test for tensor gradient. Other changes in `test_element` are formatter only.

Because we are already making fixes in the element, I thought I add value based equality too. This is rarely relevant, it would prevent recompilation of jitted functions in which an element is a static argument (i.e. as part of the operator) and you change the element instance to an equal element with the exact same quad points and values.

**Review by zrlf — APPROVED** [review]
